Fixed #30826 -- Fixed crash of many JSONField lookups when one hand side is key transform.

louiseGrandjonc opened the pull request:
When doing `as_sql` the `compiler.compile` executes this to get the sql and params.
https://github.com/django/django/blob/master/django/db/models/sql/compiler.py#L412

In the jsonb `KeyTransform` code, we return a tuple https://github.com/django/django/blob/master/django/contrib/postgres/fields/jsonb.py#L115

Which makes `PostgresSimpleLookup.as_sql` fail: https://github.com/django/django/blob/master/django/contrib/postgres/lookups.py#L11 because `lhs_params` is a tuple and `rhs_params` is a list.

```
lhs_params ('d',)
rhs_params [<django.contrib.postgres.fields.jsonb.JsonAdapter object at 0x104e7f090>]
```

The fact that it's a tuple was introduced here: https://github.com/django/django/commit/6c3dfba89215fc56fc27ef61829a6fff88be4abb to fix a subquery bug.

I have doubts on wether we should use `tuple` or `list` for params.

charettes commented on tests/postgres_tests/test_json.py:
```diff
@@ -386,6 +386,9 @@ def test_key_sql_injection(self):
             queries[0]['sql'],
         )
 
+    def test_field_contains(self):
```
I'd move this test under the existing `test_contains` method and name it `test_nested_contains`.

louiseGrandjonc replied:
Thanks, I did that. I wasn't sure if there was a rule to always add tests at the end of the `TestCase`

felixxm commented on tests/postgres_tests/test_json.py:
```diff
@@ -386,6 +386,9 @@ def test_key_sql_injection(self):
         )
 
+    def test_field_contains(self):
+        self.assertTrue(JSONModel.objects.filter(field__d__contains='e').exists())
```
Please add also a test with `icontains`
```python
self.assertTrue(JSONModel.objects.filter(field__d__icontains='e').exists())
```
I will check if other operators are also affected.

felixxm commented on tests/postgres_tests/test_json.py:
```diff
@@ -356,6 +359,9 @@ def test_iexact(self):
         self.assertFalse(JSONModel.objects.filter(field__foo__icontains='"bar"').exists())
 
+    def test_nested_icontains(self):
+        self.assertTrue(JSONModel.objects.filter(field__d__icontains='e').exists())
```
I've checked and `icontains` is not affected, we can remove this test. Sorry I should checked this first.

louiseGrandjonc replied:
So should I remove the `icontains` test? Even if it's not affected, could it be good to have a unit test on this ORM call?

felixxm replied:
This use case is already tested in `test_icontains()`.

daniel-ruiz commented:
Hi! I'm very happy to have found this PR since I'm working in a project that struggles with this same problem.

I've also experienced troubles using the `__has_key` lookup on a `JSONField`. I realized this is commented in the [original issue report](https://code.djangoproject.com/ticket/30826#comment:5), but I didn't find references to this lookup in the PR. I can help adding some regression test in case you consider it necessary.

Thanks to all for the amazing work!

felixxm commented:
@louiseGrandjonc Thanks for this patch :+1: All `JSONField`'s lookups (which inherit from `PostgresSimpleLookup`) with key transformation on the LHS or RHS are affected, i.e.
`__contains`, `__contained_by`, `__has_key`, `__has_keys`, and `__has_any_keys`. I was able to create tests for all valid combinations:
```diff
diff --git a/tests/postgres_tests/test_json.py b/tests/postgres_tests/test_json.py
index 39cba8f695..485d4060ed 100644
--- a/tests/postgres_tests/test_json.py
+++ b/tests/postgres_tests/test_json.py
@@ -135,7 +135,12 @@ class TestQuerying(PostgreSQLTestCase):
                 'k': True,
                 'l': False,
             }),
-            JSONModel(field={'foo': 'bar'}),
+            JSONModel(field={
+                'foo': 'bar',
+                'baz': {'a': 'b', 'c': 'd'},
+                'bar': ['foo', 'bar'],
+                'bax': {'foo': 'bar'},
+            }),
         ])
 
     def test_exact(self):
@@ -273,6 +278,26 @@ class TestQuerying(PostgreSQLTestCase):
             [self.objs[6], self.objs[7]]
         )
 
+    def test_key_transform_lookups(self):
+        tests = (
+            ('field__d__contains', 'e'),
+            ('field__baz__contained_by', {'a': 'b', 'c': 'd', 'e': 'f'}),
+            ('field__baz__has_key', 'c'),
+            ('field__baz__has_keys', ['a', 'c']),
+            ('field__baz__has_any_keys', ['a', 'x']),
+            ('field__contains', KeyTransform('bax', 'field')),
+            (
+                'field__contained_by',
+                KeyTransform('x', RawSQL('%s::jsonb', ['{"x": {"a": "b", "c": 1, "d": "e"}}'])),
+            ),
+            ('field__has_key', KeyTextTransform('foo', 'field')),
+        )
+        for lookup, value in tests:
+            with self.subTest(lookup=lookup):
+                self.assertTrue(JSONModel.objects.filter(
+                    **{lookup: value},
+                ).exists())
+
     def test_has_key(self):
         self.assertSequenceEqual(
             JSONModel.objects.filter(field__has_key='a'),
```
`__has_keys` and `__has_any_keys` lookups expect iterable on the RHS, so it's not possible to create filters with key transformations on the RHS for them.

Please add also release notes to the `2.2.7.txt`, `2.1.14.txt`, and `1.11.26.txt`, e.g.
```
* Fixed a crash when using a ``contains``, ``contained_by``, ``has_key``, 
  ``has_keys``, or ``has_any_keys`` lookup on
  :class:`~django.contrib.postgres.fields.JSONField`, if the right or left hand
  side of an expression is a key transformation (:ticket:`30826`).```
```


@charettes Agreed, let's move with this patch.

louiseGrandjonc commented:
> @louiseGrandjonc Thanks for this patch 👍 All `JSONField`'s lookups (which inherit from `PostgresSimpleLookup`) with key transformation on the LHS or RHS are affected, i.e.
> `__contains`, `__contained_by`, `__has_key`, `__has_keys`, and `__has_any_keys`. I was able to create tests for all valid combinations:
> 
> ```diff
> diff --git a/tests/postgres_tests/test_json.py b/tests/postgres_tests/test_json.py
> index 39cba8f695..485d4060ed 100644
> --- a/tests/postgres_tests/test_json.py
> +++ b/tests/postgres_tests/test_json.py
> @@ -135,7 +135,12 @@ class TestQuerying(PostgreSQLTestCase):
>                  'k': True,
>                  'l': False,
>              }),
> -            JSONModel(field={'foo': 'bar'}),
> +            JSONModel(field={
> +                'foo': 'bar',
> +                'baz': {'a': 'b', 'c': 'd'},
> +                'bar': ['foo', 'bar'],
> +                'bax': {'foo': 'bar'},
> +            }),
>          ])
>  
>      def test_exact(self):
> @@ -273,6 +278,26 @@ class TestQuerying(PostgreSQLTestCase):
>              [self.objs[6], self.objs[7]]
>          )
>  
> +    def test_key_transform_lookups(self):
> +        tests = (
> +            ('field__d__contains', 'e'),
> +            ('field__baz__contained_by', {'a': 'b', 'c': 'd', 'e': 'f'}),
> +            ('field__baz__has_key', 'c'),
> +            ('field__baz__has_keys', ['a', 'c']),
> +            ('field__baz__has_any_keys', ['a', 'x']),
> +            ('field__contains', KeyTransform('bax', 'field')),
> +            (
> +                'field__contained_by',
> +                KeyTransform('x', RawSQL('%s::jsonb', ['{"x": {"a": "b", "c": 1, "d": "e"}}'])),
> +            ),
> +            ('field__has_key', KeyTextTransform('foo', 'field')),
> +        )
> +        for lookup, value in tests:
> +            with self.subTest(lookup=lookup):
> +                self.assertTrue(JSONModel.objects.filter(
> +                    **{lookup: value},
> +                ).exists())
> +
>      def test_has_key(self):
>          self.assertSequenceEqual(
>              JSONModel.objects.filter(field__has_key='a'),
> ```
> 
> `__has_keys` and `__has_any_keys` lookups expect iterable on the RHS, so it's not possible to create filters with key transformations on the RHS for them.
> 
> Please add also release notes to the `2.2.7.txt`, `2.1.14.txt`, and `1.11.26.txt`, e.g.
> 
> ```
> * Fixed a crash when using a ``contains``, ``contained_by``, ``has_key``, 
>   ``has_keys``, or ``has_any_keys`` lookup on
>   :class:`~django.contrib.postgres.fields.JSONField`, if the right or left hand
>   side of an expression is a key transform (:ticket:`30826`).```
> ```
> 
> @charettes Agreed, let's move with this patch.

Thank you! I rebased with master and added the release note

felixxm commented:
@louiseGrandjonc Thanks :+1: I squashed commits and added more test cases for all affected combinations. 

Let's leave it ready to merge until next week because maybe someone will report another broken use case (that we missed).

felixxm commented:
@kmehran1106 You left a comment yesterday: _"The change to this line in 2.2.6 is breaking lookups for list in jsonfield."_. Can you provide an example? I really tried with different combinations of lookups but without any issues.
